File: src/com/dafttech/terra/game/world/entities/EntityArrow.java
package com.dafttech.terra.game.world.entities;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.entities.models.EntityThrown;
import com.dafttech.terra.game.world.tiles.Tile;
import com.dafttech.terra.game.world.tiles.TileAir;
import com.dafttech.terra.resources.Resources;

public class EntityArrow extends EntityThrown {
    Tile stuckIn = new TileAir();

    public EntityArrow(Vector2 pos, World world) {
        super(pos, world, new Vector2(2, 0.6f));

        setGravityFactor(0.25f);
    }

    @Override
    public TextureRegion getImage() {
        return Resources.ENTITIES.getImage("arrow");
    }

    @Override
    public void update(World world, float delta) {
        stuckIn = world.getTile(stuckIn.getPosition());

        if (stuckIn.isAir() || !stuckIn.isCollidableWith(this)) {
            setGravityFactor(0.25f);
        }

        super.update(world, delta);
    }
    
    @Override
    public boolean collidesWith(Entity e) {
        return !(e instanceof Player);
    }

    @Override
    public void onTerrainCollision(Tile tile) {
        stuckIn = tile;
        setGravityFactor(0);
        setVelocity(new Vector2(0, 0));
    }

    @Override
    public float getInAirFriction() {
        return 0.025f;
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/entities/EntityArrow.scala
package com.dafttech.terra.game.world.entities

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.vector.Vector2d
import com.dafttech.terra.game.world.GameWorld
import com.dafttech.terra.game.world.entities.living.Player
import com.dafttech.terra.game.world.entities.models.{EntityLiving, EntityThrown}
import com.dafttech.terra.resources.Resources
import monix.eval.Task

class EntityArrow(pos: Vector2d)(implicit world: GameWorld) extends EntityThrown(pos) {
  private[entities] var stuckIn: TilePosition = _

  override def getImage: Task[TextureRegion] = Resources.ENTITIES.getImageTask("arrow")

  override def update(delta: Float)(implicit tilePosition: TilePosition): Unit = {
    super.update(delta)
  }

  override def onTerrainCollision(tilePosition: TilePosition): Unit = {
    stuckIn = tilePosition
    //setVelocity(Vector2d(0, 0))
  }
}